### PR TITLE
[JENKINS-58110] - Added failed plugins output to failed-plugins.txt

### PR DESCRIPTION
**Pull request by sladyn98**
This PR deals with the addition of failed plugins to a new file "failed-plugins.txt".
Reference : https://issues.jenkins-ci.org/browse/JENKINS-58110

To-Do

- [ ]  Write a test case
- [ ]  Add option to change "failed-plugins.txt" path.

### CODE REVIEW

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 1]
```diff
@@ -416,6 +418,39 @@ public void checkAndSetLatestUpdateCenter() {
...
+            try {
+                String newLine = System.getProperty("line.separator");
+                FileWriter fileWriter = new FileWriter(failedPluginsFile.getPath());
+                for (Plugin plugin : failedPlugins) {
+                    String failedPluginName = plugin.getName();
+                    fileWriter.write(failedPluginName + newLine);
+                    System.out.println(failedPluginName);
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: only print the exception when `--verbose` is enabled

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 2]
```diff
@@ -416,6 +418,39 @@ public void checkAndSetLatestUpdateCenter() {
...
+        if (failedPlugins.size() > 0) {
+            System.out.println("Some plugins failed to download: ");
+            //Create a new Failed Plugins text file
+            File failedPluginsFile = new File(refDir.getAbsolutePath() + File.separator + "failed-plugins.txt");
+
+            if (!failedPluginsFile.exists()) {
+                try {
+                    failedPluginsFile.createNewFile();
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: only print the exception when `--verbose` is enabled

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 3]
```diff
@@ -416,6 +418,39 @@ public void checkAndSetLatestUpdateCenter() {
...
+                    System.out.println("File cannot be created at the path: " + failedPluginsFile
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: the message of the exception should be in this log message

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 4]
```diff
@@ -416,6 +418,39 @@ public void checkAndSetLatestUpdateCenter() {
...
         }
     }
+    /**
+     * Outputs a list of plugins which have failed to download
+     */
+    public void outputFailedPlugins() {
+        if (failedPlugins.size() > 0) {
+            System.out.println("Some plugins failed to download: ");
+            //Create a new Failed Plugins text file
+            File failedPluginsFile = new File(refDir.getAbsolutePath() + File.separator + "failed-plugins.txt");
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
            File failedPluginsFile = new File(refDir.getAbsolutePath(), "failed-plugins.txt");
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 5]
```diff
@@ -416,6 +418,39 @@ public void checkAndSetLatestUpdateCenter() {
             }
         }
     }
+    /**
+     * Outputs a list of plugins which have failed to download
+     */
+    public void outputFailedPlugins() {
+        if (failedPlugins.size() > 0) {
+            System.out.println("Some plugins failed to download: ");
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]:
```suggestion
            System.out.println(failedPlugs.size() + " plugin(s) failed to download: ");
```

File: plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java [hunk 6]
```diff
@@ -416,6 +418,39 @@ public void checkAndSetLatestUpdateCenter() {
...
+                for (Plugin plugin : failedPlugins) {
+                    String failedPluginName = plugin.getName();
+                    fileWriter.write(failedPluginName + newLine);
+                    System.out.println(failedPluginName);
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+        System.exit(1);
```
Comment (timja) [inline plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: Why? this is in the library code, we shouldn't be calling system.exit here

Comment (sladyn98) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: https://github.com/stopalopa/plugin-installation-manager-tool/blob/e31abdc5ce80e8ca95237c2c5b21fea8100a31cb/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java#L168
I picked it up from here, I thought it was essential to exiting.

Comment (timja) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: hmm k, we shouldn't be doing that, but the library needs a good amount of work to be actually usable as a library

Comment (sladyn98) [reply plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java]: okay :+1:

**Comment by sladyn98** [comment]
@timja Would we need a test for this, in which case how would we mock a failed plugin download ?

**Comment by sladyn98** [comment]
CC : @kwhetstone

**Comment by timja** [comment]
> @timja Would we need a test for this, in which case how would we mock a failed plugin download ?
> CC : @kwhetstone

Wiremock would be a good way of doing it, record a plugin being downloaded, and add a matcher for one of its dependencies and then change it to fail, possibly turn off your internet or request an invalid version and then record what happens then, depending on what you want to fail

**Comment by oleg-nenashev** [comment]
@sladyn98 hi. Do you plan to finish it? I might have a patch which will cause a merge conflict

**Comment by sladyn98** [comment]
@oleg-nenashev I do not think I have the time to complete this, because in order to implement the tests I would first have to learn how to use wiremock, am I quite overloaded on stuff to learn. So please feel free to take action as you see fit

**Comment by timja** [comment]
Wiremock is being added in https://github.com/jenkinsci/plugin-installation-manager-tool/pull/148 FTR

**Comment by sladyn98** [comment]
Closing this since no time to fix

**Pull request closed without merging**